lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,390 @@
+
+/**
+ * Existing Core Selectors
+ *
+ *     .blocks-gallery-caption
+ *     .blocks-gallery-item
+ *     .blocks-gallery-item__caption
+ *     .book-author
+ *     .message
+ *     a
+ *     a[download]
+ *     audio
+ *     blockquote
+ *     cite
+ *     code
+ *     div
+ *     figcaption
+ *     figure > a
+ *     figure a
+ *     figure img
+ *     figure video,figure img
+ *     h1,h2,h3,h4,h5,h6
+ *     img
+ *     li
+ *     ol,ul
+ *     p
+ *     pre
+ *     tbody tr
+ *     td,th
+ *     tfoot tr
+ *     thead tr
+ *     video
+ */
+
+require_once __DIR__ . '/class-wp-html-naive-processor.php';
+
+/*
+ * @see PHP docs for array_is_list: user-contributed polyfill
+ */
+if ( ! function_exists( 'array_is_list' ) ) {
+	function array_is_list( $array ) {
+		$i = 0;
+
+		foreach ( $array as $k => $v ) {
+			if ( $k !== $i++ ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+}
+
+
+class WP_HTML_Attribute_Sourcer {
+	/**
+	 * Attributes definitions, typically from `block.json`.
+	 *
+	 * @see WP_Block_Type_Registry
+	 *
+	 * @var mixed|null
+	 */
+	public $attribute_definitions;
+
+	/**
+	 * Source HTML containing embedded attributes.
+	 *
+	 * @var mixed|null
+	 */
+	private $html;
+
+	public function __construct( $attribute_definitions = null, $html = null ) {
+		$this->attribute_definitions = $attribute_definitions;
+		$this->html = $html;
+	}
+
+	public function source_attributes() {
+		$attributes = [];
+		$unparsed = [];
+
+		foreach ( $this->attribute_definitions as $name => $definition ) {
+			$sourcer = self::parse_definition( $definition );
+			switch ( $sourcer ) {
+				case null:
+				case 'not-sourced':
+				case 'unsupported':
+					$unparsed[] = $name;
+					continue 2;
+
+				case 'inner-html':
+					$attributes[ $name ] = $this->html;
+					continue 2;
+			}
+
+			$tags = self::select( $sourcer['selector'], $this->html );
+			if ( null === $tags ) {
+				$unparsed[] = $name;
+				continue;
+			}
+
+			switch ( $sourcer['type'] ) {
+				case 'html':
+					$attributes[ $name ] = self::get_inner_html( $tags );
+					continue 2;
+
+				case 'attribute':
+					$attributes[ $name ] = $tags->get_attribute( $sourcer['attribute'] );
+					continue 2;
+			}
+		}
+
+		return array(
+			'attributes' => $attributes,
+			'unparsed' => $unparsed
+		);
+	}
+
+	public static function select( $selector, $html ) {
+		$tags = new WP_HTML_Naive_Processor( $html );
+
+		if ( array_is_list( $selector ) ) {
+			while ( $tags->next_tag() ) {
+				foreach ( $selector as $s ) {
+					if ( 'element' === $s['type'] && $tags->get_tag() === strtoupper( $s['identifier'] ) ) {
+						return $tags;
+					}
+
+					// @TODO: $tags->has_class() would be _really_ handy here.
+					if ( 'class' === $s['type'] && preg_match( "~\b{$s['identifier']}\b~", $tags->get_attribute( 'class' ) ) ) {
+						return $tags;
+					}
+
+					if ( 'hash' === $s['type'] && $s['identifier'] === $tags->get_attribute( 'id' ) ) {
+						return $tags;
+					}
+				}
+			}
```

Comment:
This effectively reimplements `$tags->next_tag`. I'm tempted to say it could be included in the next_tag API as follows:

```php
$criteria = [];
foreach ( $selectors as $s ) {
    $criterion = [];
    if ( 'element' === $s['type'] ) {
      $criteria[] = [ 'tag_name' => strtoupper( $s['identifier'] ) ];
    } else if ( 'class' === $s['type'] ) {
      $criteria[] = [ 'class_name' => $s['identifier'] ];
    } else if ( 'hash' === $s['type'] ) {
      $criteria[] = [ 'id' => $s['identifier'] ];
    }
}
$tags->next_tag( [
    'or' => $criteria
] );
```

However, I can see how this may need more CSS-specific matchers that don't make sense in the core API.

At the same time, this method has duplicate logic for a special case when only one selector is received. I'd do this instead:


```php
public static function select( $selector, $html ) {
	$tags = new WP_HTML_Naive_Processor( $html );
	if ( ! array_is_list( $selector ) ) {
		$selector = [ $selector ];
	}

	while ( $tags->next_tag() ) {
		foreach ( $selector as $s ) {
			// ...
		}
	}

	return null;
}
```

Reply:
for another reason I've actually collapsed it to your suggestion; getting `has_attribute` support (such as with `a[download]`) was going to start getting cumbersome, so now it's just one big loop.